File: app/src/main/java/com/peterlaurence/trekme/ui/mapview/components/markermanage/MarkerManageFragment.java
package com.peterlaurence.trekme.ui.mapview.components.markermanage;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TextInputEditText;
import android.support.v4.app.Fragment;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.TextView;

import com.peterlaurence.trekme.R;
import com.peterlaurence.trekme.core.map.Map;
import com.peterlaurence.trekme.core.map.gson.MarkerGson;
import com.peterlaurence.trekme.core.map.maploader.MapLoader;
import com.peterlaurence.trekme.ui.MarkerProvider;
import com.peterlaurence.trekme.model.map.MapProvider;

import static android.view.View.GONE;

/**
 * A {@link Fragment} subclass that provides tools to :
 * <ul>
 * <li>Edit a marker's associated comment</li>
 * <li>See the WGS84 and projected coordinates of the marker, if possible</li>
 * <li>Delete the marker</li>
 * </ul>
 *
 * @author peterLaurence on 23/04/2017.
 */

public class MarkerManageFragment extends Fragment {
    private View rootView;
    private MarkerProvider mMarkerProvider;
    private MarkerManageFragmentInteractionListener mMarkerManageFragmentInteractionListener;

    private Map mMap;
    private MarkerGson.Marker mMarker;

    private TextInputEditText mNameEditText;
    private TextInputEditText mLatEditText;
    private TextInputEditText mLonEditText;
    private TextView mProjectionLabel;
    private TextInputEditText mProjectionX;
    private TextInputEditText mProjectionY;
    private EditText mComment;

    public interface MarkerManageFragmentInteractionListener {
        void showCurrentMap();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof MarkerProvider && context instanceof MarkerManageFragmentInteractionListener) {
            mMarkerProvider = (MarkerProvider) context;
            mMarkerManageFragmentInteractionListener = (MarkerManageFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement MapProvider and MarkerProvider");
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        rootView = inflater.inflate(R.layout.fragment_marker_manage, container, false);

        /* The view fields */
        mNameEditText = (TextInputEditText) rootView.findViewById(R.id.marker_name_id);
        mLatEditText = (TextInputEditText) rootView.findViewById(R.id.marker_lat_id);
        mLonEditText = (TextInputEditText) rootView.findViewById(R.id.marker_lon_id);
        mProjectionLabel = (TextView) rootView.findViewById(R.id.marker_proj_label_id);
        mProjectionX = (TextInputEditText) rootView.findViewById(R.id.marker_proj_x_id);
        mProjectionY = (TextInputEditText) rootView.findViewById(R.id.marker_proj_y_id);
        mComment = (EditText) rootView.findViewById(R.id.marker_comment_id);

        mMap = MapProvider.INSTANCE.getCurrentMap();
        mMarker = mMarkerProvider.getCurrentMarker();

        updateView();
        return rootView;
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        /* Hide the app title */
        ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowTitleEnabled(false);
        }

        /* Clear the existing action menu */
        menu.clear();

        /* .. and fill the new one */
        inflater.inflate(R.menu.menu_fragment_marker_manage, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.save_marker_id:
                //TODO : implement makers save to json file
                saveChanges();
                return true;
            case R.id.undo_marker_id:
                updateView();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void updateView() {
        mNameEditText.setText(mMarker.name);
        mLatEditText.setText(String.valueOf(mMarker.lat));
        mLonEditText.setText(String.valueOf(mMarker.lon));
        mComment.setText(mMarker.comment);

        /* Check whether projected coordinates fields should be shown or not */
        if (mMap.getProjection() == null) {
            mProjectionLabel.setVisibility(GONE);
            mProjectionX.setVisibility(GONE);
            mProjectionY.setVisibility(GONE);
            return;
        }

        mProjectionX.setText(String.valueOf(mMarker.proj_x));
        mProjectionY.setText(String.valueOf(mMarker.proj_y));
    }

    private void saveChanges() {
        try {
            mMarker.lat = Double.valueOf(mLatEditText.getText().toString());
            mMarker.lon = Double.valueOf(mLonEditText.getText().toString());
            mMarker.proj_x = Double.valueOf(mProjectionX.getText().toString());
            mMarker.proj_y = Double.valueOf(mProjectionY.getText().toString());
        } catch (Exception e) {
            //don't care
        }
        mMarker.name = mNameEditText.getText().toString();
        mMarker.comment = mComment.getText().toString();

        mMarkerProvider.currentMarkerEdited();

        /* Save the changes on the markers.json file */
        MapLoader.getInstance().saveMarkers(mMap);

        hideSoftKeyboard();

        /* Show a snackbar to confirm the changes and offer to show the map */
        Snackbar snackbar = Snackbar.make(rootView, R.string.marker_changes_saved, Snackbar.LENGTH_SHORT);
        snackbar.setAction(R.string.show_map_action, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mMarkerManageFragmentInteractionListener.showCurrentMap();
            }
        });
        snackbar.show();
    }

    private void hideSoftKeyboard() {
        Activity activity = getActivity();
        if (activity == null) return;

        View view = activity.getCurrentFocus();
        if (view != null) {
            InputMethodManager imm = (InputMethodManager) activity.getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }
}

File: app/src/main/java/com/peterlaurence/trekme/ui/mapview/components/markermanage/MarkerManageFragment.kt
package com.peterlaurence.trekme.ui.mapview.components.markermanage

import android.content.Context
import android.os.Bundle
import android.view.*
import android.view.View.GONE
import android.view.inputmethod.InputMethodManager
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.addTextChangedListener
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import androidx.navigation.fragment.navArgs
import com.google.android.material.snackbar.Snackbar
import com.peterlaurence.trekme.R
import com.peterlaurence.trekme.core.map.Map
import com.peterlaurence.trekme.core.map.gson.MarkerGson
import com.peterlaurence.trekme.core.map.maploader.MapLoader
import com.peterlaurence.trekme.databinding.FragmentMarkerManageBinding
import com.peterlaurence.trekme.viewmodel.markermanage.GeographicCoords
import com.peterlaurence.trekme.viewmodel.markermanage.MakerManageViewModel
import com.peterlaurence.trekme.viewmodel.markermanage.ProjectedCoords

/**
 * A [Fragment] subclass that provides tools to:
 *
 *  * Edit a marker's name
 *  * Edit a marker's associated comment
 *  * See and edit WGS84 (and projected coordinates, if applicable)
 *
 * When the latitude or longitude is changed, the projected coordinates are changed accordingly, if
 * the map has a projection. Right after this change, this do not trigger un update of lat and lon,
 * since this would result in a infinite loop.
 *
 * The reverse applies: when projected coordinates are changes, latitude and longitude are updated
 * only once.
 *
 * @author P.Laurence on 23/04/2017 -- Converted to Kotlin on 24/09/2019
 */
class MarkerManageFragment : Fragment() {
    private var _binding: FragmentMarkerManageBinding? = null
    private val binding get() = _binding!!

    private val viewModel: MakerManageViewModel by viewModels()
    private val args: MarkerManageFragmentArgs by navArgs()

    private var map: Map? = null
    private var marker: MarkerGson.Marker? = null

    /*
     * Be VERY careful not to break the logic of those flags, since their purpose is to prevent
     * infinite modification loop (geo_coord change -> proj_coord change -> geo_coord change -> ...)
     */
    private var infiniteLoopGuardGeo = false
    private var infiniteLoopGuardProj = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        viewModel.getGeographicLiveData().observe(this, Observer<GeographicCoords> {
            it?.let {
                binding.markerLonId.setText("${it.lon}")
                binding.markerLatId.setText("${it.lat}")
            }
            infiniteLoopGuardGeo = false
        })

        viewModel.getProjectedLiveData().observe(this, Observer<ProjectedCoords> {
            it?.let {
                binding.markerProjXId.setText("${it.X}")
                binding.markerProjYId.setText("${it.Y}")
            }
            infiniteLoopGuardProj = false
        })
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        _binding = FragmentMarkerManageBinding.inflate(inflater, container, false)

        map = MapLoader.getMap(args.mapId)
        map?.markers?.firstOrNull { it == args.marker }?.let {
            this.marker = it
        }

        binding.markerLatId.addTextChangedListener(
                onTextChanged = { _, _, _, _ ->
                    if (!infiniteLoopGuardGeo) {
                        onGeographicCoordsChanged()
                    }
                }
        )

        binding.markerLonId.addTextChangedListener(
                onTextChanged = { _, _, _, _ ->
                    if (!infiniteLoopGuardGeo) {
                        onGeographicCoordsChanged()
                    }
                }
        )

        binding.markerProjXId.addTextChangedListener(
                onTextChanged = { _, _, _, _ ->
                    if (!infiniteLoopGuardProj) {
                        onProjectedCoordsChanged()
                    }
                }
        )

        binding.markerProjYId.addTextChangedListener(
                onTextChanged = { _, _, _, _ ->
                    if (!infiniteLoopGuardProj) {
                        onProjectedCoordsChanged()
                    }
                }
        )

        updateView()
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        /* Hide the app title */
        val actionBar = (activity as AppCompatActivity).supportActionBar
        actionBar?.setDisplayShowTitleEnabled(false)

        /* Clear the existing action menu */
        menu.clear()

        /* .. and fill the new one */
        inflater.inflate(R.menu.menu_fragment_marker_manage, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.save_marker_id -> {
                saveChanges()
                true
            }
            R.id.undo_marker_id -> {
                updateView()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun onGeographicCoordsChanged() {
        if (infiniteLoopGuardProj) {
            return
        }
        map?.let {
            var lat: Double? = null
            var lon: Double? = null
            try {
                lat = binding.markerLatId.text?.toString()?.toDouble()
                lon = binding.markerLonId.text?.toString()?.toDouble()
            } catch (e: Exception) {
            }

            if (lat != null && lon != null) {
                infiniteLoopGuardProj = true
                viewModel.onGeographicValuesChanged(it, lat, lon)
            }
        }
    }

    private fun onProjectedCoordsChanged() {
        if (infiniteLoopGuardGeo) {
            return
        }
        map?.let {
            var X: Double? = null
            var Y: Double? = null
            try {
                X = binding.markerProjXId.text?.toString()?.toDouble()
                Y = binding.markerProjYId.text?.toString()?.toDouble()
            } catch (e: Exception) {
            }

            if (X != null && Y != null) {
                infiniteLoopGuardGeo = true
                viewModel.onProjectedCoordsChanged(it, X, Y)
            }
        }
    }

    private fun updateView() {
        val marker = this.marker ?: return
        binding.markerNameId.setText(marker.name)
        binding.markerLatId.setText(marker.lat.toString())
        binding.markerLonId.setText(marker.lon.toString())
        binding.markerCommentId.setText(marker.comment)

        /* Check whether projected coordinates fields should be shown or not */
        if (map?.projection == null) {
            binding.markerProjLabelId.visibility = GONE
            binding.markerProjXId.visibility = GONE
            binding.markerProjYId.visibility = GONE
            return
        }

        binding.markerProjXId.setText(marker.proj_x.toString())
        binding.markerProjYId.setText(marker.proj_y.toString())
    }

    private fun saveChanges() {
        val marker = this.marker ?: return
        try {
            marker.lat = java.lang.Double.valueOf(binding.markerLatId.text.toString())
            marker.lon = java.lang.Double.valueOf(binding.markerLonId.text.toString())
            marker.proj_x = java.lang.Double.valueOf(binding.markerProjXId.text.toString())
            marker.proj_y = java.lang.Double.valueOf(binding.markerProjYId.text.toString())
        } catch (e: Exception) {
            //don't care
        }

        marker.name = binding.markerNameId.text.toString()
        marker.comment = binding.markerCommentId.text.toString()

        /* Save the changes on the markers.json file */
        map?.let { MapLoader.saveMarkers(it) }

        hideSoftKeyboard()

        /* Show a snackbar to confirm the changes and offer to show the map */
        val snackbar = Snackbar.make(binding.root, R.string.marker_changes_saved, Snackbar.LENGTH_SHORT)
        snackbar.setAction(R.string.show_map_action) {
            findNavController().navigate(R.id.mapViewFragment)
        }
        snackbar.show()
    }

    private fun hideSoftKeyboard() {
        val activity = activity ?: return

        val view = activity.currentFocus
        if (view != null) {
            val imm = activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }
}